protocol/backend/eth_util.go

```diff
@@ -0,0 +1,54 @@
+package backend
+
+import (
+	"context"
+	"time"
+
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// SleepWithContext is the same as time.Sleep(...) but with preemption if the context is complete.
+func SleepWithContext(ctx context.Context, d time.Duration) {
+	timer := time.NewTimer(d)
+	select {
+	case <-ctx.Done():
+		if !timer.Stop() {
+			<-timer.C
+		}
+	case <-timer.C:
+	}
+}
+
+// WaitForEthBlockAfterTimestamp returns the header of the first block received where the block's timestamp
+// is >= ts.
+func WaitForEthBlockAfterTimestamp(ctx context.Context, ec *ethclient.Client, ts int64) (*ethtypes.Header, error) {
+	timeDelta := time.Duration(ts-time.Now().Unix()) * time.Second
+
+	// The sleep is safe even if timeDelta is negative. We only optimise for timestamps in the future, but if
+	// the timestamp had already passed for some reason, nothing bad happens other than waiting for an additional
+	// block to be mined.
```

Comment:
The last part of the comment might be wrong.  It felt like I was getting the current block header as soon as subscribing (verses not getting any headers until the next block is mined).  The documentation on this is unhelpful.

Reply:
https://pkg.go.dev/github.com/ethereum/go-ethereum/ethclient#Client.SubscribeNewHead

the docs seem to suggest it sends the current head over the channel immediately

Reply:
I adjusted the comment.  `WaitForReceipt` and `WaitForEthBlockAfterTimestamp` are now in a new package named `block` under `ethereum`.